packages/core/core-flows/src/fulfillment/steps/validate-shipping-option-prices.ts

```diff
@@ -1,26 +1,85 @@
 import { FulfillmentWorkflow } from "@medusajs/framework/types"
-import { MedusaError, Modules } from "@medusajs/framework/utils"
+import {
+  MedusaError,
+  Modules,
+  ShippingOptionPriceType,
+} from "@medusajs/framework/utils"
 import { StepResponse, createStep } from "@medusajs/framework/workflows-sdk"
 
 export const validateShippingOptionPricesStepId =
   "validate-shipping-option-prices"
 
 /**
- * Validate that regions exist for the shipping option prices.
+ * Validate that shipping options can be crated based on provided price configuration.
+ *
+ * For flat rate prices, it validates that regions exist for the shipping option prices.
+ * For calculated prices, it validates with the fulfillment provider if the price can be calculated.
  */
 export const validateShippingOptionPricesStep = createStep(
   validateShippingOptionPricesStepId,
   async (
-    options: {
-      prices?: FulfillmentWorkflow.UpdateShippingOptionsWorkflowInput["prices"]
-    }[],
+    options: (
+      | FulfillmentWorkflow.CreateShippingOptionsWorkflowInput
+      | FulfillmentWorkflow.UpdateShippingOptionsWorkflowInput
+    )[],
     { container }
   ) => {
-    const allPrices = options.flatMap((option) => option.prices ?? [])
+    const fulfillmentModuleService = container.resolve(Modules.FULFILLMENT)
+
+    const optionIds = options.map(
+      (option) =>
+        (option as FulfillmentWorkflow.UpdateShippingOptionsWorkflowInput).id
+    )
+
+    if (optionIds.length) {
+      /**
+       * This means we are validating an update of shipping options.
+       * We need to ensure that all shipping options have price_type set
+       * to correctly determine price updates.
+       *
+       * (On create, price_type must be defined already.)
+       */
+      const shippingOptions =
+        await fulfillmentModuleService.listShippingOptions(
+          {
+            id: optionIds,
+          },
+          { select: ["id", "price_type", "provider_id"] }
+        )
+
+      const optionsMap = new Map(
+        shippingOptions.map((option) => [option.id, option])
+      )
+
+      ;(
+        options as FulfillmentWorkflow.UpdateShippingOptionsWorkflowInput[]
+      ).forEach((option) => {
+        option.price_type =
+          option.price_type ?? optionsMap.get(option.id)?.price_type
+        option.provider_id =
+          option.provider_id ?? optionsMap.get(option.id)?.provider_id
+      })
+    }
+
+    const flatRatePrices = options.flatMap((option) =>
+      option.price_type === ShippingOptionPriceType.FLAT
+        ? ((option.prices ?? []) as { region_id: string; amount: number }[])
+        : []
+    )
+
+    const calculatedOptions = options
+      .filter(
+        (option) => option.price_type === ShippingOptionPriceType.CALCULATED
+      )
+      .flatMap((option) => option ?? [])
```

Comment:
nit(non-blocking): lots of looping over the same options here – perhaps we could simply:
```
const flatRatePrices = []
const calculatedOptions = []

options.forEach((option) => {
  option.price_type = option.price_type ?? optionsMap.get(option.id)?.price_type
  option.provider_id = option.provider_id ?? optionsMap.get(option.id)?.provider_id

  if (option.price_type === ShippingOptionPriceType.FLAT) {
    flatRatePrices.push(...(option.prices ?? []))
  }

  if (option.price_type === ShippingOptionPriceType.CALCULATED) {
    calculatedOptions.push(option)
  }
})
```